vpr/SRC/route/route_tree_timing.c

```diff
@@ -479,16 +562,600 @@ void free_route_tree(t_rt_node * rt_node) {
+					else rt_node->re_expand = true;
+					rt_node->parent_node = parent_node;
+					rt_node->parent_switch = prev_edge->iswitch;
+
+					// add to dangling edge (switch determined when edge was created)
+					prev_edge->child = rt_node;
+					// at this point parent and child should be set and mutual
+					INCREMENTAL_REROUTING_ASSERT(parent_node->u.child_list->child == rt_node);
+					INCREMENTAL_REROUTING_ASSERT(rt_node->parent_node == parent_node);
+					// create new dangling edge belonging to the current node for its child (there exists one since new_node isn't tail (sink))
+					prev_edge = alloc_linked_rt_edge();
+					prev_edge->iswitch = new_node->iswitch;
+					prev_edge->next = nullptr;
+					rt_node->u.child_list = prev_edge;
+					// hold onto rt_node since we will need it when it becomes the join point for another branch
+					rr_node_to_rt_node[inode] = rt_node;
+					parent_node = rt_node;
+				}
+				// still need to add the sink to the route tree
+				t_rt_node* rt_node = alloc_rt_node();
+				rt_node->u.child_list = nullptr;	// no children for sink (edge is not allocated, need to do that outside for next branch)
+				rt_node->inode = tail->index;
+				rt_node->re_expand = false;	// do not re_expand SINKS
+				rt_node->parent_switch = prev_edge->iswitch;
+				rt_node->parent_node = parent_node;
+				prev_edge->child = rt_node;
+
+				// at this point parent and child should be set and mutual
+				INCREMENTAL_REROUTING_ASSERT(parent_node->u.child_list->child == rt_node);
+				INCREMENTAL_REROUTING_ASSERT(rt_node->parent_node == parent_node);
+
+				rr_node_to_rt_node[tail->index] = rt_node;
+				// terminates this series of edges/branch
+				prev_edge = nullptr;
+				break;
+			}
+		}
+		// either another prexisting part of the route tree and more sinks to come or end
+		head = tail->next;
+		// end of trunk
+		if (!head) return rt_root;
+		// continue on the next branch (1 after a sink for the traceback data structure)
+		// the rt_node has already been alocated as it's part of the trunk, make a new dangling edge for it
+		parent_node = rr_node_to_rt_node[head->index];
+		prev_edge = alloc_linked_rt_edge();
+		prev_edge->iswitch = head->iswitch;
+		// advance to trunk node's next edge
+		prev_edge->next = parent_node->u.child_list;
+		parent_node->u.child_list = prev_edge;
+		// edge is still dangling in preparation for next sink
+	}
+}
+
+static t_trace* traceback_branch_from_route_tree(t_trace* head, const t_rt_node* root, int& sinks_left) {
+
+	/* Transforms route tree (supposed to be a branch of the full tree) into  *
+	 * a traceback segement 												  *
+	 * returns the **sink trace (tail)** 									  *
+	 * assumes head has been allocated with index set, but no iswitch or next */
+
+	t_linked_rt_edge* edge {root->u.child_list};
+	
+	for (;;) {
+		// head has been allocated at the start of this loop and has correct index
+
+		// reached a sink
+		if (!edge) {
+			head->iswitch = OPEN;
+			--sinks_left;
```

Comment:
This is by reference, so it modifies the callers sinks_left.  Would be good to explicitly state this.